server/server.go

```diff
@@ -38,10 +38,25 @@ func NewServer(w lookout.Watcher, p lookout.Poster, fileGetter lookout.FileGette
 
 // Run starts server
 func (s *Server) Run(ctx context.Context) error {
-	// FIXME(max): we most probably want to change interface of EventHandler instead of it
-	return s.watcher.Watch(ctx, func(e lookout.Event) error {
-		return s.handleEvent(ctx, e)
-	})
+	errCh := make(chan error, 1)
+	go func() {
+		// FIXME(max): we most probably want to change interface of EventHandler instead of it
+		err := s.watcher.Watch(ctx, func(e lookout.Event) error {
+			return s.handleEvent(ctx, e)
+		})
+		errCh <- err
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-errCh:
+		if err != nil {
+			log.Errorf(err, "error during watching, restart watching")
+			return s.Run(ctx)
```

Comment:
Could you help me understand - would this be an equivalent of the infinite retry policy, in case of networking failure, though a recursive `s.Run()` calls?

Reply:
It's equivalent of the infinite retry policy in case of any failure in the watcher. (except manual stopping by the special type of error, most common error here is a network of course but might be something else)

It means server wouldn't die if something went wrong there (dying at least for now isn't a very good option. We can post comments at the same time for example)

From the administrative point of view, it means we should monitory error logs, not only server restarts.

Reply:
makes sense!

What confuses me is not infinite retry policy per se, but potentially infinite recursion, as heap usually is not infinite.